File: 백준/Gold/BOJ_1916.java
package gold;

import java.sql.Array;
import java.util.*;
import java.io.*;

public class BOJ_1916 {
    static int N, M;
    static List<Node>[] graph;
    static int[] dist;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());
        M = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++)
            graph[i] = new ArrayList<>();

        graph = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int S = Integer.parseInt(st.nextToken()) - 1;
            int E = Integer.parseInt(st.nextToken()) - 1;
            int cost = Integer.parseInt(st.nextToken());
            if (graph[S].get(E).cost > cost)
                graph[S].add(new Node(E, cost));



        }


    }



    public static class Node {
        int target;
        int cost;

        public Node(int curr, int cost) {
            this.target = target;
            this.cost = cost;
        }
    }
}

File: 백준/Gold/BOJ_1916.py
import heapq
import sys

def bfs(start):
    hq = []
    heapq.heappush(hq, (0, start))
    dist[start] = 0
    while hq:
        cost, curr = heapq.heappop(hq)

        if (cost > dist[curr]):
            continue

        for target in graph[curr]:
            next, new_cost = target
            if dist[next] > cost + new_cost:
                dist[next] = cost + new_cost
                heapq.heappush(hq, (new_cost, next))


N = int(input())
M = int(input())
dist = [float('inf')] * N
graph = [[] for _ in range(N)]
for i in range(M):
    S, E, cost = map(int, sys.stdin.readline().split())
    graph[S-1].append((E-1, cost));

start, end = map(int, sys.stdin.readline.split())
bfs(start)
print(dist[end])



